File: cpp_100-999/122.cc
// C++
// leetcode 122
// https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock-ii/
// 

class Solution {
public:
    int maxProfit(vector<int>& prices) {
    	int pre = prices[0];
    	int nMinusOne = prices.size() - 1;
    	int res = 0;
    	for(int i = 1; i < prices.size(); i++) {
    		if(prices[i] > pre)
    			res += prices[i] - pre;
    		pre = prices[i];
    	}
    	return	res;
    }
};

File: Go/Go_100-999/122.go
// Go
// leetcode 122
// https://leetcode.cn/problems/best-time-to-buy-and-sell-stock-ii/

func maxProfit(prices []int) int {
	res := 0
	takeMax := -1000000
	for _, x := range prices {
		next := max(res, takeMax + x)
		takeMax = max(takeMax, res - x)
		res = next
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}